### PR TITLE
[1/N] Initial vllm-ext evaluation support (MXFP4 MOE)

### PR DESCRIPTION
**Pull request by yiliu30**
### Usage
```bash
source apply_ext.sh 
VLLM_ENABLE_AR_EXT=1 vllm serve ...
```

Initial support for out-of-tree AutoRound integration with vLLM, verified on `Qwen3-15B-A2B-Base`. 
- [x] Added `AutoRoundExtensionConfig`
- [x] Monkey patching support
- [x] Added `AutoRoundMoEMethod` and `AutoRoundMoEMethodMXFp4Impl`
- [x] UT

There are three paths for MXFP4
- `VLLM_ENABLE_STATIC_MOE=1`: static MOE + unpack the weight on the fly, very slow, for correctness check
- `VLLM_ENABLE_STATIC_MOE=1` + `VLLM_MXFP4_PRE_UNPACK_WEIGHTS=1`, static MOE + unpack the weight to FP8 before inference, still slow but acceptable
- `VLLM_AR_MXFP4_MODULAR_MOE=1` + `VLLM_MXFP4_PRE_UNPACK_WEIGHTS=1`, unpack the weight to BF16 + FP32 group gemm, fast but requires the same mem as BF16 models, WIP, support it in the next PR

### CODE REVIEW

File: auto_round/experimental/vllm_ext/fp4_utils.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+import torch
+
+FLOAT_TO_E2M1 = [
```
Comment (mengniwang95) [inline auto_round/experimental/vllm_ext/fp4_utils.py]: it is not used

Comment (yiliu30) [reply auto_round/experimental/vllm_ext/fp4_utils.py]: removed

File: auto_round_extension/vllm_ext/__init__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+# ==---------------------------------------------------------------------------==
+# Apply the extension
+# ==---------------------------------------------------------------------------==
+
+
+def apply():
+    import vllm.model_executor.layers.quantization.auto_round as auto_round_module
+
+    from .auto_round_ext import AutoRoundExtensionConfig
```
Comment (n1ck-guo) [inline auto_round_extension/vllm_ext/__init__.py]: better use absolute import

Comment (yiliu30) [reply auto_round_extension/vllm_ext/__init__.py]: Updated

File: auto_round_extension/vllm_ext/utils.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        data_hp: High precision data.
+        max_abs: Maximum absolute value for data_hp along specified dimension/block_size.
+        max_pos: The maximum value of the low precision data type.
+
+    Returns:
+        exponent: The biased exponent with dtype E8M0 in uint8 container.
+        data_lp: The targeted low precision data, in high precision container
+            (requires cast to low precision data type).
+    """
+    descale = max_abs / max_pos
```
Comment (n1ck-guo) [inline auto_round_extension/vllm_ext/utils.py:60]: Is it necessary to determine whether the value is reasonable here?

File: auto_round_extension/vllm_ext/quant_method_moe.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+import torch
+from vllm.logger import init_logger
+from vllm.model_executor.layers.fused_moe import (
+    FusedMoEConfig,
+    FusedMoEMethodBase,
+)
+from vllm.model_executor.layers.fused_moe.config import FusedMoEQuantConfig
+from vllm.model_executor.layers.quantization.auto_round import AutoRoundConfig
+
```
Comment (wenhuach21) [inline auto_round_extension/vllm_ext/quant_method_moe.py:25]: better add a readme in this folder to tell users/teammates about how to use this feature

File: auto_round_extension/vllm_ext/moe_impl_mxfp4.py [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+        IN = intermediate_size_per_partition
+        if self.has_bias:
+            # TODO: yiliu30 use the dtype in CK
+            bias_dtype = torch.bfloat16
+            w13_bias = torch.nn.Parameter(torch.zeros(E, 2 * IN, dtype=bias_dtype), requires_grad=False)
+            layer.register_parameter("w13_bias", w13_bias)
+            set_weight_attrs(w13_bias, extra_weight_attrs)
+
+            w2_bias = torch.nn.Parameter(
+                torch.zeros(num_experts, hidden_size, dtype=bias_dtype),
```
Comment (mengniwang95) [inline auto_round_extension/vllm_ext/moe_impl_mxfp4.py]: why not use E and H

File: auto_round_extension/vllm_ext/fp4_utils.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    # Apply sign bit (bit 3) to get final 4-bit representation
+    indices = abs_indices + (torch.signbit(x).to(torch.long) << 3)
+
+    # Reshape to prepare for packing pairs of values
+    indices = indices.reshape(-1)
+
+    # Handle odd length by padding if necessary
+    if indices.numel() % 2 != 0:
+        indices = torch.cat([indices, torch.zeros(1, dtype=torch.long, device=device)])
```
Comment (mengniwang95) [inline auto_round_extension/vllm_ext/fp4_utils.py]: For odd length, where do you remove the padding 0?

File: auto_round_extension/vllm_ext/utils.py (hunk 3 above)
Comment (yiliu30) [reply auto_round_extension/vllm_ext/utils.py:60]: The `max_pos` is maximum value of the low precision data type, which is 6 in this case.

File: auto_round_extension/vllm_ext/fp4_utils.py (hunk 6 above)
Comment (yiliu30) [reply auto_round_extension/vllm_ext/fp4_utils.py]: Updated, thx!

File: auto_round_extension/vllm_ext/moe_impl_mxfp4.py (hunk 5 above)
Comment (yiliu30) [reply auto_round_extension/vllm_ext/moe_impl_mxfp4.py]: Updated

File: auto_round_extension/vllm_ext/quant_method_moe.py (hunk 4 above)
Comment (yiliu30) [reply auto_round_extension/vllm_ext/quant_method_moe.py:25]: Sure, added usage in PR description, will add formal docs in the later PR.

File: auto_round_extension/vllm_ext/moe_impl_mxfp4.py [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+                self.mask_weights_buffer = torch.zeros(
+                    (num_all_tokens, total_num_experts), dtype=x.dtype, device=x.device
+                )
+                self.experts_mask_buffer = torch.zeros(
+                    (num_all_tokens, total_num_experts), dtype=x.dtype, device=x.device
+                )
+                mask_weights = self.mask_weights_buffer
+                experts_mask = self.experts_mask_buffer
+        else:
+            self.mask_weights_buffer.zero_()
```
Comment (mengniwang95) [inline auto_round_extension/vllm_ext/moe_impl_mxfp4.py:409]: self.mask_weights_buffer maybe None?

Comment (yiliu30) [reply auto_round_extension/vllm_ext/moe_impl_mxfp4.py:409]: The `mask_weights_buffer` was checked at L379.
